### PR TITLE
Token support

### PR DESCRIPTION
**Pull request by lavrin**

### CODE REVIEW

File: src/escalus_session.erl [hunk 1]
```diff
@@ -317,3 +327,12 @@ get_stream_management(Features) ->
 -spec get_advanced_message_processing(exml:element()) -> boolean().
 get_advanced_message_processing(Features) ->
     undefined =/= exml_query:subelement(Features, <<"amp">>).
+
+get_sasl_mechanisms_info(Features) ->
+    OAUTH = exml_query:subelement(Features, <<"mechanisms">>),
+    case OAUTH of
+        #xmlel{} ->
+            #xmlel{children = SaslMechs} = OAUTH,
+            lists:map(fun(X) -> {binary_to_atom(exml_query:cdata(X), latin1), true} end, SaslMechs);
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 336 is too long:             lists:map(fun(X) -> {binary_to_atom(exml_query:cdata(X), latin1), true} end, SaslMechs);.

File: src/escalus_session.erl [hunk 2]
```diff
@@ -292,11 +295,18 @@ assert_stream_features(StreamFeatures, Transport, IsLegacy) ->
 
 -spec get_stream_features(exml:element()) -> features().
 get_stream_features(Features) ->
-    [{compression, get_compression(Features)},
-     {starttls, get_starttls(Features)},
-     {stream_management, get_stream_management(Features)},
-     {advanced_message_processing, get_advanced_message_processing(Features)}
-    ].
+    ResultFeatures =
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 298 has 1 trailing whitespace characters.

File: src/escalus_session.erl [hunk 3]
```diff
@@ -292,11 +295,18 @@ assert_stream_features(StreamFeatures, Transport, IsLegacy) ->
...
-    [{compression, get_compression(Features)},
-     {starttls, get_starttls(Features)},
-     {stream_management, get_stream_management(Features)},
-     {advanced_message_processing, get_advanced_message_processing(Features)}
-    ].
+    ResultFeatures = 
+        [{compression, get_compression(Features)},
+         {starttls, get_starttls(Features)},
+         {stream_management, get_stream_management(Features)},
+         {advanced_message_processing, get_advanced_message_processing(Features)}
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 302 is too long:          {advanced_message_processing, get_advanced_message_processing(Features)}.

File: src/escalus_session.erl [hunk 4]
```diff
@@ -83,9 +83,12 @@ starttls(Conn, Props) ->
 authenticate(Conn, Props) ->
     %% FIXME: as default, select authentication scheme based on stream features
     {M, F} = proplists:get_value(auth, Props, {escalus_auth, auth_plain}),
-    M:F(Conn, Props),
+    PropsAfterAuth = case M:F(Conn, Props) of
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 86. Only modules that define callbacks should make dynamic calls.

File: src/escalus_session.erl (hunk 3 above)
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 302 is too long:          {advanced_message_processing, get_advanced_message_processing(Features)}.

File: src/escalus_session.erl (hunk 4 above)
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 86. Only modules that define callbacks should make dynamic calls.

File: src/escalus_session.erl (hunk 1 above)
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 336 is too long:             lists:map(fun(X) -> {binary_to_atom(exml_query:cdata(X), latin1), true} end, SaslMechs);.

File: src/escalus_auth.erl [hunk 5]
```diff
@@ -87,6 +88,21 @@ auth_sasl_external(Conn, Props) ->
...
 
+auth_sasl_oauth(Conn, Props) ->
+    Token = get_property(oauth_token, Props),
+    Children = [#xmlcdata{content = base64:encode(Token)}],
+    Stanza = escalus_stanza:auth(<<"X-OAUTH">>, Children),
+    ok = escalus_connection:send(Conn, Stanza),
+    AuthReply = escalus_connection:get_stanza(Conn, auth_reply),
+    NewProps = case AuthReply of
+                   #xmlel{name = <<"success">>, children = ChildrenRecvd} ->
+                       ([ {oauth_returned_token, base64:decode(exml:unescape_cdata(CData))}
```
Comment (gadgetci) [inline src/escalus_auth.erl]: According to **Elvis**:
> Line 99 is too long:                        ([ {oauth_returned_token, base64:decode(exml:unescape_cdata(CData))}.

File: src/escalus_session.erl (hunk 2 above)
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 298 has 1 trailing whitespace characters.

File: src/escalus_auth.erl (hunk 5 above)
Comment (gadgetci) [inline src/escalus_auth.erl]: According to **Elvis**:
> Line 99 is too long:                        ([ {oauth_returned_token, base64:decode(exml:unescape_cdata(CData))}.

File: src/escalus_session.erl [hunk 6]
```diff
@@ -83,9 +83,12 @@ starttls(Conn, Props) ->
 authenticate(Conn, Props) ->
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Compiler**:
> Warning: missing specification for function authenticate/2

File: src/escalus_auth.erl [hunk 7]
```diff
@@ -87,6 +88,21 @@ auth_sasl_external(Conn, Props) ->
     ok = escalus_connection:send(Conn, Stanza),
     wait_for_success(ThisServer, Conn).
 
+auth_sasl_oauth(Conn, Props) ->
```
Comment (gadgetci) [inline src/escalus_auth.erl]: According to **Compiler**:
> Warning: missing specification for function auth_sasl_oauth/2

Comment (gadgetci) [inline src/escalus_auth.erl]: According to **Compiler**:
> Warning: missing specification for function auth_sasl_oauth/2

File: src/escalus_session.erl (hunk 6 above)
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Compiler**:
> Warning: missing specification for function authenticate/2

File: src/escalus_session.erl [hunk 8]
```diff
@@ -317,3 +320,12 @@ get_stream_management(Features) ->
 -spec get_advanced_message_processing(exml:element()) -> boolean().
 get_advanced_message_processing(Features) ->
     undefined =/= exml_query:subelement(Features, <<"amp">>).
+
+-spec get_sasl_mechanisms(exml:element()) -> features().
+get_sasl_mechanisms(Features) ->
+    MechCDataPath = [{element, <<"mechanisms">>}, {element, <<"mechanism">>}, cdata],
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Line 326 is too long:     MechCDataPath = [{element, <<"mechanisms">>}, {element, <<"mechanism">>}, cdata],.

File: src/escalus_session.erl [hunk 9]
```diff
@@ -77,18 +80,24 @@ start_stream(Conn, Props) ->
...
+-spec starttls(client(), user_spec()) -> {client(), user_spec()}.
 starttls(Conn, Props) ->
     escalus_tcp:upgrade_to_tls(Conn, Props).
 
+-spec authenticate(client(), user_spec()) -> user_spec().
 authenticate(Conn, Props) ->
     %% FIXME: as default, select authentication scheme based on stream features
     {M, F} = proplists:get_value(auth, Props, {escalus_auth, auth_plain}),
-    M:F(Conn, Props),
+    PropsAfterAuth = case M:F(Conn, Props) of
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 91. Only modules that define callbacks should make dynamic calls.

Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 91. Only modules that define callbacks should make dynamic calls.

File: src/escalus_auth.erl (hunk 7 above)
Comment (gadgetci) [inline src/escalus_auth.erl]: According to **Compiler**:
> Warning: missing specification for function auth_sasl_oauth/2

File: src/escalus_session.erl [hunk 10]
```diff
@@ -77,18 +87,24 @@ start_stream(Conn, Props) ->
...
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 98. Only modules that define callbacks should make dynamic calls.

**Comment by lavrin** [comment]
I left the _dynamic function call_ warning/Elvis setting on purpose, but don't intend to fix it in this PR. Unless there's something missing, this is ready to merge.

**Comment by lavrin** [comment]
BTW, this builds on top of #92.

File: src/escalus_session.erl [hunk 11]
```diff
@@ -317,3 +336,15 @@ get_stream_management(Features) ->
...
+-spec get_sasl_mechanisms(exml:element()) -> features().
+get_sasl_mechanisms(Features) ->
+    MechCDataPath = [{element, <<"mechanisms">>}, {element, <<"mechanism">>}, cdata],
+    %% TODO: The expected modules are stored as a custom attribute,
+    %%       so they don't get optimized out by the compiler.
+    %%       By calling `?MODULE:module_info()` we force them to be read into
+    %%       the ERTS atom table from the attributes chunk.
+    ?MODULE:module_info(),
+    %% TODO: convert these to escalus_auth function names
+    [ {binary_to_existing_atom(Mech, utf8), true}
```
Comment (michalwski) [inline src/escalus_session.erl]: Do we really need atoms here? Can't we just use binaries?

Comment (lavrin) [inline src/escalus_session.erl]: TBH, I don't know where it's used. The plan was to leave it in a state which is good to go forward without changing the interface (i.e. format of `Features` returned), but I still have some time left so will look for places actually using this code.

File: src/escalus_session.erl [hunk 12]
```diff
@@ -77,18 +88,24 @@ start_stream(Conn, Props) ->
...
+-spec starttls(client(), user_spec()) -> {client(), user_spec()}.
 starttls(Conn, Props) ->
     escalus_tcp:upgrade_to_tls(Conn, Props).
 
+-spec authenticate(client(), user_spec()) -> user_spec().
 authenticate(Conn, Props) ->
     %% FIXME: as default, select authentication scheme based on stream features
     {M, F} = proplists:get_value(auth, Props, {escalus_auth, auth_plain}),
-    M:F(Conn, Props),
+    PropsAfterAuth = case M:F(Conn, Props) of
```
Comment (gadgetci) [inline src/escalus_session.erl]: According to **Elvis**:
> Remove the dynamic function call on line 99. Only modules that define callbacks should make dynamic calls.

**Pull request merged**